routing.go

```diff
@@ -93,39 +97,55 @@ func (dht *IpfsDHT) GetValue(ctx context.Context, key string) (_ []byte, err err
 	ctx, cancel := context.WithTimeout(ctx, time.Minute)
 	defer cancel()
 
-	vals, err := dht.GetValues(ctx, key, 16)
-	if err != nil {
+	var cfg ropts.Options
+	if err := cfg.Apply(opts...); err != nil {
 		return nil, err
 	}
 
-	recs := make([][]byte, 0, len(vals))
-	for _, v := range vals {
-		if v.Val != nil {
-			recs = append(recs, v.Val)
-		}
-	}
-	if len(recs) == 0 {
-		return nil, routing.ErrNotFound
-	}
-
-	i, err := dht.Selector.BestRecord(key, recs)
+	results, err := dht.getValues(ctx, key, &cfg)
 	if err != nil {
 		return nil, err
 	}
 
-	best := recs[i]
-	log.Debugf("GetValue %v %v", key, best)
-	if best == nil {
-		log.Errorf("GetValue yielded correct record with nil value.")
-		return nil, routing.ErrNotFound
+	var outdatedPeers, currentPeers []peer.ID
+
+	var best []byte
+	for result := range results {
+		switch {
+		case result.val == nil:
+			outdatedPeers = append(outdatedPeers, result.from)
+		case best == nil:
+			best = result.val
+			fallthrough
+		case bytes.Equal(result.val, best):
+			currentPeers = append(currentPeers, result.from)
+		default:
+			i, err := dht.Selector.BestRecord(key, [][]byte{best, result.val})
+			if err != nil {
+				log.Error(err)
+				return nil, err
+			}
+			switch i {
+			case 0:
+				outdatedPeers = append(outdatedPeers, result.from)
+			case 1:
+				outdatedPeers = append(outdatedPeers, currentPeers...)
+				currentPeers = append(currentPeers[:0], result.from)
```

Comment:
I am a little confused by this -- what's the semantics of `i` in the return value?

Reply:
I can add some more documentation to the definition in the go-libp2p-record package (`SelectorFunc` is documented but selectors themselves are not).

Reply:
The documentation is nice and useful, but it would be good to have a line here explaining the reasoning (and not having the reader dig the documentation for selector).

Reply:
Copying function documentation to call sites seems a bit redundant (most editors *do* have goto definition support) but I've documented the cases a bit.

Reply:
Right, didn't want you to copy the function documentation, just clarify what the case switch does!